### PR TITLE
Test/ml 353 misuse of etch ml model

### PR DESCRIPTION
**Pull request by iryna-melnyk**

### CODE REVIEW

File: libs/vm-modules/tests/unit/ml/ml_model_tests.cpp [hunk 1]
```diff
@@ -848,4 +848,140 @@ TEST_F(VMModelTests, model_with_accuracy_metric)
...
+          model.fit(data, label, 32u64);
+
+          // evaluate
+          var mets = model.evaluate();
+          return mets;
+        endfunction
+      )";
+
+  ASSERT_TRUE(toolkit.Compile(SRC_METRIC));
+  ASSERT_FALSE(toolkit.Run());
```
Comment (baykaner) [inline libs/vm-modules/tests/unit/ml/ml_model_tests.cpp:879]: great! 👍  this behaviour is consistent with tensorflow

File: libs/vm-modules/tests/unit/ml/ml_model_tests.cpp [hunk 2]
```diff
@@ -848,4 +848,140 @@ TEST_F(VMModelTests, model_with_accuracy_metric)
...
+}
+
+TEST_F(VMModelTests, model_sequential_multiple_compile)
+{
+  static char const *SEQUENTIAL_SRC = R"(
+        function main()
+          var model = Model("sequential");
+          model.add("dense", 10u64, 10u64, "relu");
+          model.compile("mse", "adam");
+          model.compile("scel", "adam");
```
Comment (baykaner) [inline libs/vm-modules/tests/unit/ml/ml_model_tests.cpp:902]: it would be good if we could verify somehow that model.compile is actually changing the loss function for us each time.

Also, I would suspect that model will be adding a new op each time we call compile, but maybe it doesn't delete the old loss op from the previous compile call.

Can you investigate these two points please?

File: libs/vm-modules/tests/unit/ml/ml_model_tests.cpp [hunk 3]
```diff
@@ -848,4 +848,140 @@ TEST_F(VMModelTests, model_with_accuracy_metric)
...
+}
+
+TEST_F(VMModelTests, model_sequential_add_after_compile)
+{
+  static char const *SEQUENTIAL_SRC = R"(
+      function main()
+         var model = Model("sequential");
+         model.add("dense", 10u64, 10u64, "relu");
+         model.compile("mse", "adam");
+         model.add("dense", 10u64, 1u64, "relu");
```
Comment (baykaner) [inline libs/vm-modules/tests/unit/ml/ml_model_tests.cpp:918]: again - let's verify that after the second call to compile we have a 2 layer network, and that we have a 1 layer network before that

File: libs/vm-modules/tests/unit/ml/ml_model_tests.cpp [hunk 4]
```diff
@@ -848,4 +848,140 @@ TEST_F(VMModelTests, model_with_accuracy_metric)
...
+
+        return mets;
+      endfunction
+    )";
+
+  ASSERT_TRUE(toolkit.Compile(SEQUENTIAL_SRC));
+  ASSERT_TRUE(toolkit.Run());
+}
+
+TEST_F(VMModelTests, DISABLED_model_sequential_predict_bad_data)
```
Comment (baykaner) [inline libs/vm-modules/tests/unit/ml/ml_model_tests.cpp:964]: lets enable this test but EXPECT_THROW:

seems reasonable to throw when you are passing a tensor of size 0 in for prediction

File: libs/vm-modules/tests/unit/ml/ml_model_tests.cpp [hunk 5]
```diff
@@ -848,4 +848,140 @@ TEST_F(VMModelTests, model_with_accuracy_metric)
...
+        model.compile("mse", "adam", {"categorical accuracy"});
+
+        var prediction = model.predict(data);
+      endfunction
+    )";
+
+  ASSERT_TRUE(toolkit.Compile(SEQUENTIAL_SRC));
+  ASSERT_FALSE(toolkit.Run());
+}
+
```
Comment (baykaner) [inline libs/vm-modules/tests/unit/ml/ml_model_tests.cpp:986]: please add another test to check EXPECT_THROW if we compile and then evaluate (this should not be permitted without training first)

File: libs/vm-modules/tests/unit/ml/ml_model_tests.cpp (hunk 4 above)
Comment (iryna-melnyk) [reply libs/vm-modules/tests/unit/ml/ml_model_tests.cpp:964]: I've found a valid statement for this case - EXPECT_DEATH(toolkit.Run(), "1000");

File: libs/vm-modules/tests/unit/ml/ml_model_tests.cpp (hunk 5 above)
Comment (iryna-melnyk) [reply libs/vm-modules/tests/unit/ml/ml_model_tests.cpp:986]: Done

File: libs/vm-modules/tests/unit/ml/ml_model_tests.cpp (hunk 2 above)
Comment (iryna-melnyk) [reply libs/vm-modules/tests/unit/ml/ml_model_tests.cpp:902]: So, each time when model.compile(...) is called we add to graph an error node:
graph_ptr_->template AddNode<ops::CrossEntropyLoss<TensorType>>("Error", {output_, label_});)
when we try to add to graph a node with existing name it checks for duplicates and returns an exception:
ml::exceptions::InvalidMode( "OperationType is not shareable. Cannot make duplicate of node named: " + node_name);

File: libs/vm-modules/tests/unit/ml/ml_model_tests.cpp (hunk 3 above)
Comment (iryna-melnyk) [reply libs/vm-modules/tests/unit/ml/ml_model_tests.cpp:918]: True. When compile is called second time we really have a 2-layered network because node names are unique for each dense layer. But again, it throws an InvalidMode exception because of ambiguous node names ("Error" node already exists in the graph)

File: libs/vm-modules/tests/unit/ml/ml_model_tests.cpp (hunk 2 above)
Comment (baykaner) [reply libs/vm-modules/tests/unit/ml/ml_model_tests.cpp:902]: so when is this actually failing? On the second Compile call? (with a different loss op "scel"), or on the third compile call (which reuses "mse")?

File: libs/vm-modules/tests/unit/ml/ml_model_tests.cpp (hunk 3 above)
Comment (baykaner) [reply libs/vm-modules/tests/unit/ml/ml_model_tests.cpp:918]: hmmm.... okay, well I feel like the ideal behaviour would be have .add trigger a 'decompile' - which handles removing the node from the graph (but we have no infrastructure to support that currently) - so I think it would be a tricky job

File: libs/vm-modules/tests/unit/ml/ml_model_tests.cpp (hunk 2 above)
Comment (iryna-melnyk) [reply libs/vm-modules/tests/unit/ml/ml_model_tests.cpp:902]: On the second Compile call? - yes

File: libs/vm-modules/tests/unit/ml/ml_model_tests.cpp (hunk 3 above)
Comment (iryna-melnyk) [reply libs/vm-modules/tests/unit/ml/ml_model_tests.cpp:918]: I can create a ticket for that

**Comment by iryna-melnyk** [comment]
look at https://github.com/fetchai/ledger/pull/2171
